cmd/erasure-readfile.go

```diff
@@ -17,328 +17,142 @@
-	for block := startBlock; block <= endBlock; block++ {
-		// Mark all buffers as unused at the start of the loop so that the buffers
-		// can be reused.
+	blocks := make([][]byte, len(s.disks))
+	for off := offset / blocksize; length > 0; off++ {
+		blockOffset := off * chunksize
 		pool.Reset()
 
-		// Each element of enBlocks holds curChunkSize'd amount of data read from its corresponding disk.
-		enBlocks := make([][]byte, len(disks))
-
-		if ((offset + bytesWritten) / blockSize) == (totalLength / blockSize) {
-			// This is the last block for which curBlockSize and curChunkSize can change.
-			// For ex. if totalLength is 15M and blockSize is 10MB, curBlockSize for
-			// the last block should be 5MB.
-			curBlockSize = totalLength % blockSize
-			curChunkSize = getChunkSize(curBlockSize, dataBlocks)
+		if currentBlock := (offset + f.Size) / blocksize; currentBlock == lastBlock {
+			blocksize = totalLength % blocksize
+			chunksize = getChunkSize(blocksize, (s.dataBlocks+s.parityBlocks)/2)
 		}
-
-		// NOTE: That for the offset calculation we have to use chunkSize and
-		// not curChunkSize. If we use curChunkSize for offset calculation
-		// then it can result in wrong offset for the last block.
-		blockOffset := block * chunkSize
-
-		// nextIndex - index from which next set of parallel reads
-		// should happen.
-		nextIndex := 0
-
-		for {
-			// readDisks - disks from which we need to read in parallel.
-			var readDisks []StorageAPI
-			var err error
-			// get readable disks slice from which we can read parallelly.
-			readDisks, nextIndex, err = getReadDisks(disks, nextIndex, dataBlocks)
-			if err != nil {
-				return bytesWritten, err
-			}
-			// Issue a parallel read across the disks specified in readDisks.
-			parallelRead(volume, path, readDisks, disks, enBlocks, blockOffset, curChunkSize, brVerifiers, pool)
-			if isSuccessDecodeBlocks(enBlocks, dataBlocks) {
-				// If enough blocks are available to do rs.Reconstruct()
-				break
-			}
-			if nextIndex == len(disks) {
-				// No more disks to read from.
-				return bytesWritten, traceError(errXLReadQuorum)
-			}
-			// We do not have enough enough data blocks to reconstruct the data
-			// hence continue the for-loop till we have enough data blocks.
+		err = s.readConcurrent(volume, path, blockOffset, chunksize, blocks, verifiers, locks, pool)
+		if err != nil {
+			return f, traceError(errXLReadQuorum)
 		}
 
-		// If we have all the data blocks no need to decode, continue to write.
-		if !isSuccessDataBlocks(enBlocks, dataBlocks) {
-			// Reconstruct the missing data blocks.
-			if err := decodeData(enBlocks, dataBlocks, parityBlocks); err != nil {
-				return bytesWritten, err
-			}
+		writeLength := blocksize - startOffset
+		if length < writeLength {
+			writeLength = length
 		}
+		n, err := writeDataBlocks(writer, blocks, (s.dataBlocks+s.parityBlocks)/2, startOffset, writeLength)
```

Comment:
We should be using implicit read quorum of `(s.dataBlocks+s.parityBlocks)/2` here
(applies multiple times in this function)